flexsite/settings_local_template.txt
"""
Custom local settings for your flexelog installion.

Can be generated by `python flexelog_setup.py`, then edited

For more information on Django settings, see
https://docs.djangoproject.com/en/5.2/topics/settings/
and https://docs.djangoproject.com/en/5.2/ref/settings/
"""
import sys
from pathlib import Path
from .base_settings import *

# -------------------
# MANDATORY SETTINGS 
# You must set these for your local installation
# -------------------
# Generate a secret key at a command line with:
# python -c "from django.core.management.utils import get_random_secret_key; print(get_random_secret_key())"
SECRET_KEY = "{SECRET_KEY}"

# Path to where you want to store the flexelog database
# If running completely independent groups of logbooks, use separate "Top Groups"
# Perhaps e.g. "sales", "engineering" etc.
ELOG_DIR = r"{ELOG_DIR}"
TOP_GROUP = r"{TOP_GROUP_NAME}"
DATABASE_NAME = "{DATABASE_NAME}"

MEDIA_ROOT = Path(ELOG_DIR) / TOP_GROUP / "media"
DELETED_MEDIA = Path(ELOG_DIR) / TOP_GROUP / "deleted_media"

# If user deletes attachment, whether to leave it, delete it, or move to DELETED_MEDIA
# If false, can run a script to check for any files not in the Attachment table
ON_ATTACHMENT_DELETE = "Move"  # or "Delete" or "Leave" 

# TIME ZONE / LANGUAGE
TIME_ZONE = "{TIME_ZONE}"
USE_TZ = True  # Recommend True, dates still displayed in local timezone.
LANGUAGE_CODE = "en-us"
USE_I18N = True  # Recommend True, gives translations where available


# Database
# https://docs.djangoproject.com/en/5.1/ref/settings/#databases

DATABASES = {{
    "default": {{                                                                                                                                    
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": Path(ELOG_DIR) / TOP_GROUP / DATABASE_NAME,
    }}
}}


# -----------------------------
# SETTINGS TO OPTIONALLY CHANGE IF NEEDED 
# -----------------------------
# AUTH SETTINGS
PASSWORD_RESET_TIMEOUT = 60*15  # seconds user has to use a password reset link
SESSION_EXPIRE_AT_BROWSER_CLOSE = False
SESSION_COOKIE_AGE = 28 * 60 * 60 * 24  # 60*60*24 = one day


# Names of Groups auto-created for a new logbook
# Users are not automatically assigned to those groups
_view = ("view_entries",)
_own = ("add_entries", "edit_own_entries", "delete_own_entries",)
_others = ("edit_others_entries", "delete_others_entries")
_admin = ("configure_logbook",)

DEFAULT_LOGBOOK_GROUP_PERMISSIONS = {{
    "Logbook {{logbook.name}} Viewers": _view,
    "Logbook {{logbook.name}} Contributors": _view + _own,
    "Logbook {{logbook.name}} Editors": _view + _own + _others,
    "Logbook {{logbook.name}} Admin": _view + _own + _others + _admin,
}}

# ATTACHMENTS:
FILE_UPLOAD_MAX_SIZE = 104_857_600  # 100 MiB


# When adding new in Admin webpages, starting values for config
# just to give some examples to edit
GLOBAL_CONFIG_INITIAL = """\
Reverse sort = True  # most recent at top
Display mode = summary,  # default mode for search
Entries per page = 20

# When listing entries, the columns shown
#   '*attributes' means all Attributes not explicitly listed otherwise
#   You can override this in each logbook config, for specific logbook Attributes
List display = ID, Date, Author, *attributes, Text, Attachments 

# Approximate size of Text summary when listing entries
Summary lines = 3
Summary line length= 100,
"""

LOGBOOK_CONFIG_INITIAL = """\
Attributes = Project, Category, Subject
Required attributes = Project, Subject
Options Project = Flexelog, ClimateChange, Politics
MOptions Category = Communications, Documentation, Meetings, General

Entries per page = 30
Preset on first reply Subject = Re: $Subject
"""

flexsite/settings.py
# Copyright 2025 flexelog authors. See LICENSE file for details.
"""
Django settings for flexsite project.

Generated by 'django-admin startproject' using Django 5.1.4.

For more information on this file, see
https://docs.djangoproject.com/en/5.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/5.1/ref/settings/
"""

from pathlib import Path
import sys
from django.utils.translation import gettext_lazy as _
from django.urls import reverse_lazy

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent
RUNNING_TESTS = "test" in sys.argv

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/5.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "django-insecure-^xsftv8x!jvc++0_pv+5#4-aaz1220)hm!8mtdzn%c%uh&e*^("

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True  # XXXXX

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = [
    "flexelog.apps.FlexelogConfig",
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "guardian",
    "django.contrib.humanize",
    "django_htmx",
]

if DEBUG and not RUNNING_TESTS:
    INSTALLED_APPS.append("debug_toolbar")

MIDDLEWARE = [ 
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.locale.LocaleMiddleware",  # Remove if have only fixed locale
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "django_htmx.middleware.HtmxMiddleware",
]

if DEBUG and not RUNNING_TESTS:
    MIDDLEWARE.insert(0, "debug_toolbar.middleware.DebugToolbarMiddleware")

X_FRAME_OPTIONS = "SAMEORIGIN"  # allow viewing attachments in iframe

ROOT_URLCONF = "flexsite.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
                "flexelog.elog_cfg.cfg_context",
            ],
        },
    },
]

WSGI_APPLICATION = "flexsite.wsgi.application"
# Database
# https://docs.djangoproject.com/en/5.1/ref/settings/#databases

DATABASES = {
    "default": {                                                                                                                                    
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": BASE_DIR / "flexelog demo.db",
    }
}


AUTHENTICATION_BACKENDS = (
    'django.contrib.auth.backends.ModelBackend',  # this is default
    'guardian.backends.ObjectPermissionBackend',
)


# Password validation
# https://docs.djangoproject.com/en/5.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]
PASSWORD_RESET_TIMEOUT = 60*15  # seconds
LOGIN_URL = reverse_lazy('login')
LOGIN_REDIRECT_URL = "/"
LOGOUT_REDIRECT_URL = reverse_lazy('login')

INTERNAL_IPS = [
    "127.0.0.1",
]

# Internationalization
# https://docs.djangoproject.com/en/5.1/topics/i18n/

LANGUAGES = [
    ('en', _("English")),
    ('bg', _("Bulgarian")),
    ('br', _("Brazilian")),
    ('cz', _("Czech")),
    ('de', _("German")),
    ('dk', _("Danish")),
    ('es', _("Spanish")),
    ('fr', _("French")),
    ('id', _("Indonesia")),
    ('it', _("Italian")),
    ('jp', _("Japanese")),
    ('nl', _("Dutch")),
    ('pl', _("Polish")),
    ('ru', _("Russian")),
    ('se', _("Swedish")),
    ('sk', _("Slowak")),
    ('tr', _("Turkish")),
    ('zh', _("Chinese")),
]

LOCALE_PATHS = [
    BASE_DIR / "flexelog" / "locale",
]

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/5.1/howto/static-files/

STATIC_URL = "static/"

# Default primary key field type
# https://docs.djangoproject.com/en/5.1/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

if DEBUG:
    MEDIA_ROOT = BASE_DIR / "media"
    MEDIA_URL = "media/"

FILE_UPLOAD_MAX_SIZE = 10485760  # 10 MiB

# USER OVERRIDES / ADDITIONS imported in separate file
#   so not in version control
try:
    from flexsite.settings_local import *
except ImportError as e:
    raise Exception(
        "Please supply a settings_local.py file in the flexsite folder\n"
        "Run `python flexlog_setup.py` or copy `settings_local_template.py` and edit it."
    ) from e
